src/System.Text.Json/src/System/Text/Json/JsonDocument.cs

```diff
@@ -0,0 +1,436 @@
+        {
+            if (_utf8Json.IsEmpty)
+                throw new ObjectDisposedException(nameof(JsonDocument));
+
+            _parsedData.Get(index, out DbRow row);
+
+            if (row.TokenType != JsonTokenType.Number)
+            {
+                throw new InvalidOperationException();
+            }
+
+            ReadOnlySpan<byte> data = _utf8Json.Span;
+            ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
+
+            if (Utf8Parser.TryParse(segment, out int tmp, out int consumed) &&
+                consumed == segment.Length)
+            {
+                value = tmp;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal bool TryGetValue(int index, out long value)
+        {
+            if (_utf8Json.IsEmpty)
+                throw new ObjectDisposedException(nameof(JsonDocument));
+
+            _parsedData.Get(index, out DbRow row);
+
+            if (row.TokenType != JsonTokenType.Number)
+            {
+                throw new InvalidOperationException();
+            }
+
+            ReadOnlySpan<byte> data = _utf8Json.Span;
+            ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
+
+            if (Utf8Parser.TryParse(segment, out long tmp, out int consumed) &&
+                consumed == segment.Length)
+            {
+                value = tmp;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal bool TryGetValue(int index, out ulong value)
+        {
+            if (_utf8Json.IsEmpty)
+                throw new ObjectDisposedException(nameof(JsonDocument));
+
+            _parsedData.Get(index, out DbRow row);
+
+            if (row.TokenType != JsonTokenType.Number)
+            {
+                throw new InvalidOperationException();
+            }
+
+            ReadOnlySpan<byte> data = _utf8Json.Span;
+            ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
+
+            if (Utf8Parser.TryParse(segment, out ulong tmp, out int consumed) &&
+                consumed == segment.Length)
+            {
+                value = tmp;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        internal bool TryGetValue(int index, out double value)
+        {
+            if (_utf8Json.IsEmpty)
+                throw new ObjectDisposedException(nameof(JsonDocument));
+
+            _parsedData.Get(index, out DbRow row);
+
+            if (row.TokenType != JsonTokenType.Number)
+            {
+                throw new InvalidOperationException();
+            }
+
+            ReadOnlySpan<byte> data = _utf8Json.Span;
+            ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
+
+            if (Utf8JsonReader.TryGetDoubleValue(segment, out double tmp))
```

Comment:
The integral types have a simple representation in JSON that aligns with .NET.

Double has either dotted decimal or scientific notation, and JsonReader already has to understand which format it has so that it can inform Utf8Parser what the parse format is, so I exposed that to piggy back off of it.